cli/src/main/java/com/devonfw/tools/ide/cli/IdeCompleter.java

```diff
@@ -0,0 +1,217 @@
+package com.devonfw.tools.ide.cli;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import org.jline.reader.Candidate;
+import org.jline.reader.Completer;
+import org.jline.reader.LineReader;
+import org.jline.reader.ParsedLine;
+import org.jline.reader.impl.completer.ArgumentCompleter;
+import org.jline.reader.impl.completer.NullCompleter;
+import org.jline.utils.AttributedString;
+
+import com.devonfw.tools.ide.commandlet.Commandlet;
+import com.devonfw.tools.ide.commandlet.ContextCommandlet;
+import com.devonfw.tools.ide.commandlet.HelpCommandlet;
+import com.devonfw.tools.ide.context.IdeContext;
+import com.devonfw.tools.ide.property.FlagProperty;
+import com.devonfw.tools.ide.property.LocaleProperty;
+import com.devonfw.tools.ide.property.Property;
+import com.devonfw.tools.ide.property.ToolProperty;
+import com.devonfw.tools.ide.property.VersionProperty;
+import com.devonfw.tools.ide.tool.ToolCommandlet;
+import com.devonfw.tools.ide.version.VersionIdentifier;
+
+/**
+ * Implements the {@link Completer} for jline3 autocompletion. Inspired by picocli
+ */
+public class IdeCompleter extends ArgumentCompleter implements Completer {
+
+  /** Checks for invalid pattern e.g. '- foo foobar' */
+  private static final String INVALID_PATTERN_PART1 = "([\\-](\\s)+.*$";
+
+  /** Checks for invalid pattern e.g. ' foo - ' or ' - foo' */
+  private static final String INVALID_PATTERN_PART2 = "([\\s][a-z]*[\\s][\\-](\\s)*.*$)";
+
+  /** Checks for invalid pattern e.g. 'get-version foobar - foobar' */
+  private static final String INVALID_PATTERN_PART3 = "([a-z][\\-][a-z]*[\\s][\\-](\\s)*.*$))";
+
+  /** Checks for invalid pattern e.g. 'foo-bar - foobar' */
+  private static final String INVALID_PATTERN_PART4 = "(^[a-z]*(\\s)[\\-](\\s)*.*$)";
+
+  /** Pattern which should stop autocompletion e.g. 'get-version mvn -' (prevents options after tool commands) */
+  private static final String INVALID_PATTERN = INVALID_PATTERN_PART1 + "|" + INVALID_PATTERN_PART2 + "|" + INVALID_PATTERN_PART3
+  + "|" + INVALID_PATTERN_PART4;
```

Comment:
If I get it right what is checked here altogether is that a hyphen may not occur in isolation.
* Can't this be done with `(^|[\\s])-($|[\\s])` much simpler and much more efficient?
* What is the impact if the user does this? I know it is an edge-case but we already support GNU `--` option in our CLI parser (that allows to provide a file named `-` or `-f` without parsing it as a CLI option). At least for commands like `rm` this is important as if you by accident create a file named `-` or `-f` you could never delete it again :)

Reply:
Thanks for your suggestion but this is more complicated I think.
See behavior of the autocompletion tests when I use your suggestion:
```
expected: "-t -f get-version -"
 but was: "-t -f get-version get-version"
```
and:
```
expected: "get-version -t "
 but was: "get-version -t az"
```
The autocompleter should stop when we can't provide any valid options.